### ISSUE TITLE
Access Token request does not handle client secret basic well

### ISSUE THREAD

**Issue by claweyenuk** [OpenIDC/pyoidc#543]
I am not 100% sure if this bug is on my end or the lib.  I am using a slightly modified version of the simple examples, which I have working on my machine with the implicit hybrid flow.  I changed to using the code flow, and the request to get the access token is failing.

I put some debug in `src/oic/utils/authn/client.py` :
```
def verify_client(inst, areq, authn, type_method=TYPE_METHOD):
    """
    Initiated Guessing !

    :param inst: Entity instance
    :param areq: The request
    :param authn: client authentication information
    :return: tuple containing client id and client authentication method
    """
    print("CHECKING", authn, areq, type(areq))
```

And I get:
```
CHECKING None {'grant_type': 'authorization_code', 'code': 'Z0FBQUFBQmJCMGJiZVdWeHh3VlJLZm9CVGI3eU9HUl8yUmUyenZ4Y0xRR
GgxeXY3cFlZVlZESnJETWFUcGhHa3RobDd3S3hxd01sNHFNSUxpa3FUbkRmZURhTEZYV2UzU1J5eVNiU3JKUlYyeW5aWkVOTDVTX3NBNVRPWDJEWFlneE
5ZU2RrY0JqclZDSmhPT0JSRjVqbmdIazJ4bDl4S1BrYzhBbFBWNHhPNGU0N2RscVJZMURWa2tuT0J4OFFPb0lIZ1E5dWxlZ1Z0NlhkTjZ3Vk5YbVl0ZS1
DWWhGNmZ0bGEzYVZjTmEzOGFQdWt0bVdBTGdhWT0=', 'redirect_uri': 'https://localhost:8003/code_flow', 'state': 'Bf1xxC1D66ZqDOV0'} <class 'oic.oic.message.AccessTokenRequest'>
```

For some reason `authn` is `None`, and because this is client secret basic, the client secret is not passed along, so the code falls to `"Missing client authentication."` and errors.  I *believe* the None authn is why this is failing, but I cannot trace back why that is unset.

The headers the OP server sees are:

```
{'ACTUAL_SERVER_PROTOCOL': 'HTTP/1.1', 'PATH_INFO': '', 'QUERY_STRING': '', 'REMOTE_ADDR': '127.0.0.1', '
REMOTE_PORT': '30668', 'REQUEST_METHOD': 'POST', 'REQUEST_URI': b'/token', 'SCRIPT_NAME': '/token', 'SERVER_NAME': 'N
eon', 'SERVER_PROTOCOL': 'HTTP/1.1', 'SERVER_SOFTWARE': 'CherryPy/3.2.4 Server', 'wsgi.errors': <_io.TextIOWrapper na
me='<stderr>' mode='w' encoding='utf-8'>, 'wsgi.input': <cherrypy.wsgiserver.wsgiserver3.KnownLengthRFile object at 0
x000001CED07632E8>, 'wsgi.multiprocess': False, 'wsgi.multithread': True, 'wsgi.run_once': False, 'wsgi.url_scheme':
'https', 'wsgi.version': (1, 0), 'SERVER_PORT': '443', 'HTTP_HOST': 'localhost', 'HTTP_USER_AGENT': 'python-requests/
2.18.4', 'HTTP_ACCEPT_ENCODING': 'gzip, deflate', 'HTTP_ACCEPT': '*/*', 'HTTP_CONNECTION': 'keep-alive', 'HTTP_AUTHOR
IZATION': 'Basic dVJzQUI1T3I1ZUM4OjE4ZDY0NjQ2MDg4ZWY5YTQ5OWUyYjBhMDFiZjBhYmJhMTliYjkzM2QwNGY4MWM3YjBmZDE1Mjhj', 'CONT
ENT_TYPE': 'application/x-www-form-urlencoded', 'CONTENT_LENGTH': '471', 'HTTPS': 'on', 'SSL_PROTOCOL': 'TLSv1/SSLv3'
, 'SSL_CIPHER': 'ECDHE-RSA-AES256-GCM-SHA384'}
```

It is possible that the registration is bad.  The response there I also printed out:

```
 {'client_id': 'uRsAB5Or5eC8', 'client_secret': '18d64646088ef9a499e2b0a01bf0abba19bb933d04f81c7b0fd1528c', 'registration_access_token': 'dVdKuD6IXot8B7bUG85dJsoMZ3Lu5ZJK', 'registration_client_uri': 'https://localhost/registration?client_id=uRsAB5Or5eC8', 'client_secret_expires_at': 1527289942, 'client_id_issued_at': 1527203542, 'application_type': 'web', 'response_types': ['code'], 'request_uris': ['requests/7939af4c0f1ebe4049e933a07a667d0f58c0529cad7478808e6fabaec343492b'], 'grant_types': ['authorization_code'], 'redirect_uris': ['https://localhost:8003/code_flow']}
```

This doesn't include a `"token_endpoint_auth_method"` field so the code sets that to a default of `client_secret_basic`.  I am not sure if that is correct, and that could be the real bug.

**Comment by tpazderka** [comment]
I am not sure about the environment you are using, but the `AUTHORIZATION` header seems to be correctly filled. It looks like it is not passed to the code.

**Comment by tpazderka** [comment]
Closed by #542
